src/cpp/shared/game/game.cpp

```diff
@@ -268,130 +267,124 @@ void game::ProcessOnMapHeroes() {
-            FREE(ppMapExtra[ppMapExtraHeroIdx]);
-            ppMapExtra[ppMapExtraHeroIdx] = 0;
           }
+        }
+        for (artifactIdx = 0; artifactIdx < 3; ++artifactIdx) {
+          if (mapExtraHero->artifacts[artifactIdx] >= 0) {
+            GiveArtifact(randomHero, mapExtraHero->artifacts[artifactIdx], 1, -1);
+          }
+        }
+        if (mapExtraHero->relatedToName[5]) {
+          strcpy(randomHero->name, &mapExtraHero->relatedToName[6]);
+        }
+        randomHero->experience = 0;
+        gpAdvManager->GiveExperience(randomHero, mapExtraHero->experience, 1);// field_17 changed to experience
+        randomHero->CheckLevel();       // Check this function (design question based on generalizing hardcoded, faction-specific data structure information related to skills)
+        randomHero->x = coordXForRandomHero;
+        randomHero->y = coordYForRandomHero;
+
+        if (isJail) {
+          randomHero->ownerIdx = -1;
+          this->relatedToHeroForHireStatus[mapExtraHero->heroID] = 65;
+        } else {
+          randomHero->ownerIdx = mapExtraHero->owner;
+          this->relatedToHeroForHireStatus[mapExtraHero->heroID] = randomHero->ownerIdx;
+          this->players[randomHero->ownerIdx].heroesOwned[this->players[randomHero->ownerIdx].numHeroes++] = randomHero->idx;
+        }
+        if (!isJail && coordYForRandomHero > 0 && (this->map.tiles[coordXForRandomHero + ((coordYForRandomHero - 1) * this->map.width)].objType) == 163) {
+          --randomHero->relatedToY;
+          --randomHero->y;
+          this->castles[this->GetTownId(coordXForRandomHero, coordYForRandomHero - 1)].visitingHeroIdx = randomHero->idx;
+        }
+        if (isJail) {
+          loc->extraInfo = mapExtraHero->heroID;
+        } else {
+          loc->objTileset = 0;
+          loc->objectIndex = -1;
+          loc->extraInfo = 0;
+          loc->objType = 0;
+        }
+
+        if (mapExtraHero->couldBeHasSecondarySkills) {
+          randomHero->numSecSkillsKnown = 0;
+          for (secondarySkillIdxA = 0; secondarySkillIdxA < 14; ++secondarySkillIdxA) { //clear secondary skills
+            randomHero->secondarySkillLevel[secondarySkillIdxA] = 0;
+            randomHero->skillIndex[secondarySkillIdxA] = 0;
+          }
+          for (secondarySkillIdxB = 0; secondarySkillIdxB < 8; ++secondarySkillIdxB) { //assign new secondary skills
+            if (mapExtraHero->secondarySkills[secondarySkillIdxB] != -1)
+              randomHero->GiveSS(mapExtraHero->secondarySkills[secondarySkillIdxB],  *(&mapExtraHero->firstSecondarySkillLevel + secondarySkillIdxB));
+          }
+        }
 
+        if (!isJail) {
+          this->SetVisibility(randomHero->x, randomHero->y, randomHero->ownerIdx, giVisRange[randomHero->secondarySkillLevel[3]]);
         }
+
+        FREE(ppMapExtra[ppMapExtraHeroIdx]);
+        ppMapExtra[ppMapExtraHeroIdx] = 0;
       }
-    }
   }
+}
```

Comment:
Uhhh.....what? Does this compile? I don't think these braces match up.

Reply:
The last change committed should have both compiled and functioned properly. I have just checked, and it does.

Reply:
Although, I thought I had fixed the spacing issue there. It's fixed now...for now...